Build openzwave locally

julienw opened the pull request.

julienw commented:
I'm not sure this works properly on MacOS X, I'd need somebody to try locally.

julienw commented on openzwave-sys/openzwave-c/manager.h:
```diff
@@ -1,5 +1,5 @@
-#include <openzwave/Manager.h>
-#include <openzwave/Driver.h>
+#include <Manager.h>
+#include <Driver.h>
```
this is necessary because the open-zwave source directory does not have a openzwave directory :(

julienw commented on openzwave-sys/build.rs:
```diff
@@ -6,10 +31,29 @@ fn main() {
+    let this_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+
+    // then tell rust to link with open-zwave
+    println!("cargo:rustc-link-lib=udev");
```
we need -ludev, I don't know how to make this more automatic, and I don't think this works on mac :(

dhylands replied:
Yeah udev only exists on linux. So you'll probably need to have two separate function to do the println! (one for linux and one one for non-linux) and use:
#[cfg(target_os = "linux")] and #[cfg(not(target_os = "linux"))] in front of the 2 functions.

julienw commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "openzwave-sys/open-zwave"]
+	path = openzwave-sys/open-zwave
+	url = https://github.com/OpenZWave/open-zwave.git
```
I hope cargo automatically updates the submodules when cloning...

dhylands replied:
What's the motivations for using a sub-module?

julienw replied:
This is super easy to update (and this is done like this in some similar rust crates (eg https://github.com/alexcrichton/git2-rs/tree/master/libgit2-sys, https://github.com/carllerche/curl-rust/tree/master/curl-sys) -- some other crates are copying though)

dhylands replied:
FYI: cargo doesn't seem to update submodules (at least not for local builds).
I had to do git submodule init && git submodule update to get the build to work.

dhylands replied:
It does seem to do the right thing when the link to openwave is a github repo

julienw commented on openzwave-sys/build.rs:
```diff
@@ -1,4 +1,64 @@
+        .arg("-thin").arg("x86_64")
+        .arg("-output").arg("libopenzwave-thin.a")
+        .arg("libopenzwave.a")
+        .status()
```
this is ugly but this is the best way I found given I can't easily generate a thin library from open-zwave Makefile :(

dhylands replied:
Why does it matter? Won't the linker just pick the bits that it wants? I'm not a Mac developer, so this is more a question than a comment.

julienw replied:
no, this doesn't work :( rustc has no support for fat binaries

Before I did this it didn't build

julienw replied:
Fat binaries is a thing of the past, that's why rust developers doesn't want to support it. We can see what will happen on openzwave's side. I somehow hope that they will remove the build for x86.

julienw commented on openzwave-sys/build.rs:
```diff
@@ -6,10 +66,28 @@ fn main() {
-    println!("cargo:rustc-flags=-l openzwave");
+     .flag("-std=c++11"); // to iterate with ranges
+
+    if find_include_path(&mut c).is_ok() {
```
I'll reuse hub's patch to use pkg-config if it's found, in a separate PR.

julienw replied:
likely we'll also want to skip it if we're cross compiling -- but that's for a separate PR too.

dhylands replied:
We shouldn't use pkg-config if we're building open-zwave as a sub-module. We know exactly where everything is.

julienw commented on openzwave-sys/build.rs:
```diff
@@ -6,10 +66,28 @@ fn main() {
+    if find_include_path(&mut c).is_ok() {
+        c.compile("libopenzwave-c.a");
+        println!("cargo:rustc-flags=-l openzwave");
+        return;
```
this is the previous behavior

julienw commented:
@dhylands I think this is ready to be reviewed

julienw commented:
I tested on MacOS X without openzwave, and on a linux computer with/without openzwave installed. But I wouldn't mind a double check.

dhylands commented on openzwave-sys/build.rs:
```diff
@@ -1,4 +1,29 @@
+    // TODO try to actually build with these paths
+    let possible_paths = vec![
+        "/usr/include/openzwave",
+        "/use/local/include/openzwave",
```
If we're building from a sub-module, we shouldn't require the headers to be installed as a dependency, We should be using the headers from the sub-module.

r- for this

julienw replied:
the logic is that only if I don't find those I build using the local submodule; otherwise I use the installed lib.

julienw commented:
ping @dhylands, this is ready for a second review:
- removed the build using the system library
- refactored how "make" is called and added "-j4"
- add a line to .gitmodules to ignore changes to the submodule, so that we don't have it shown in `git status` after a `cargo build`.

dhylands commented:
r+

I was able to build. I confirmed that its linking open-zwave statically and that it linked the locally built version and not my installed version.
